src/common/formattable.h

```diff
@@ -62,14 +60,105 @@ struct to_string_formatter : fmt::formatter<std::string_view> {
     }
 };
 
-template <typename T>
-struct to_hex_string_formatter : fmt::formatter<std::string_view> {
-    template <typename FormatContext>
-    auto format(const T& val, FormatContext& ctx) const {
-        if constexpr (::formattable::detail::callable_to_hex_string_method<T>)
-            return formatter<std::string_view>::format(val.to_hex_string(), ctx);
-        else
-            return formatter<std::string_view>::format(to_hex_string(val), ctx);
+// fmt-compatible formatter for things like pubkeys that have a .data() and .size(); if printed bare
+// (i.e. "{}") then these get output as `<00000000000000000123456789abcdef>`, but we also permit
+// them to be formatted using a format of `{:x}`, `{:z}`, `{:a}`, `{:b}`, or `{:r}` where `x` means
+// the full hex byte value, `z` means the hex value without leading 0's, `:a` is base32z, `:b` is
+// base64, and `:r` means output raw bytes.  If no format string is given at all (i.e. just "{}")
+// then the format is equivalent to `{:x}`, i.e. we return the full hex value.  Other flags (such as
+// alignment, fill, width) are not currently supported.
+//
+// For example, for input span \0\0\0\0\0\0\0\x02 `fmt::format("0x{:z} 0x{:x} {}", val)` would
+// produced the string `"0x2 0x0000000000000002 0000000000000002"`.
+//
+// Some types extend this to override the default format -- for example, crypto::eth_address's
```

Comment:
This footnote with special behaviour for `eth_address` seems a little bit tricky. It's a special case that's makes the API behave less predictable. TBF I believe this is only used at the fringe of the codebase, e.g. when displaying to the user, maybe at JSON serialisation due and ETH APIs requiring strict `0x` prefixes but it's in those cases where I'd want to be very explicit about adding a '0x' prefix or not.